uPy/html.py

```diff
@@ -0,0 +1,255 @@
+            # Create a new dictionary to store info
+            tagDict={}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("\ttagDict[\"tag_type\"]={0}".format(tag_internals[0]))
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key]=val
+                #print("\ttagDict[\"{0}\"]={1}".format(key,val))
+            
+            #append this tag into the form
+            if(formOpen):
+                    inside_list=form["inside"]   #get
+                    inside_list.append(tagDict)  #append
+                    form["inside"]=inside_list   #put back
+
+    return form
+
+
+# TOP LEVEL FUNCTION
+# Input: HTML splashpage as bytes
+# Returns a list of dictionaries. Each dictionary is a form:    
+#   Pick a form by indexing into the list. Each form is a dictionary.   
+#   Dictionary key value pairs are keys and values within that tag line:
+#        ex from html: name="formname" method="POST"
+#        becomes:   dict["name"] -> "formname"
+#                   dict["method"] -> "POST"
+def get_forms(html):
+    
+    html = html.decode('utf-8') # int to str
+
+    # each entry in list is "<form ...> <...> </form>" 
+    forms = get_tags(html, "form")
+    if(len(forms) == 0):
+        return []
+    elif(len(forms) == 1): #caused if <form...> but bad html received no </form>
+        forms = []      #get rid of previous bad entry
+        ret_val = find_complete_string(html, "<form", "</div", 0) #WORKAROUND: try going until the end of the next div
+        if(ret_val[0] == ""):
+            return []   #gah lee you've just hit the worst written splashpage ever. just give up already
+        forms.append(ret_val[0])
+    
+    #debugging
+    print("All forms as complete strings:")
+    i=0
+    while i < len(forms):
+        print( "form[{0}]=\n{1}".format(i, forms[i]) )
+        i+=1
+
+
+    #break forms as strings into a list of tags: [form] = ["<form ...>", "<...>", "</form>"]
+    all_forms = []
+    for string in forms:
+        print("Before pulling out tags: {0}".format(string))
+        ret = get_tags(string)
+        all_forms.append(ret)
+        print("After pulling out tags: {0}".format(ret))
+
+    #debugging
+    print("\nEach form broken into a list of strings:")
+    i=0
+    while i < len(all_forms):
+        print( "form[{0}]=".format(i) )
+        for tag in all_forms[i]:
+            print(tag)
+        i+=1
+
+    # split out each tag's internals
+    print("\nSplitting up tag internals:")
+    i = 0
+    while i < len(all_forms):
+        tags = all_forms[i]
+        j = 0
+        while j < len(tags):
+            tag = tags[j]
+            print("Tag before breakup: {0}".format(tag))
+            all_forms[i][j] = breakup_tag(tag)
+            print("After Breakup Form[{0}] tag[{1}]: {2}".format(i, j, all_forms[i][j]) )
```

Comment:
before and after tag prints, just comment the structure of the data no need for these prints